### PR TITLE
Upgrade build to sbt 1.1.0

### PR DESCRIPTION
**Pull request by SystemFw**
Initial attempt to update the shapeless build to sbt 1.0, currently blocked by several plugins. I will add more details later to help track progress on this.

### CODE REVIEW

File: project/build.properties [hunk 1]
```diff
@@ -1 +1 @@
-sbt.version=0.13.15
+sbt.version=1.0.0-RC2
```
Comment (dwijnand) [inline project/build.properties]: RC3 is out.

Comment (SystemFw) [inline project/build.properties]: Yes, it doesn't seem to be on brew yet and I knew this was blocked anyway for now, hence the laziness :)
I'll make sure everything is up to date by the time I finish

Comment (dwijnand) [inline project/build.properties]: sbt's launcher (as installed by brew or others) is intended to be forward-compatible with sbt.version's defined in build.properties, fyi.

Comment (SystemFw) [inline project/build.properties]: Thanks for all the help with sbt stuff these days! :)

Comment (dwijnand) [inline project/build.properties]: It's what I (am paid to) do. :D

File: project/plugins.sbt [hunk 2]
```diff
@@ -4,12 +4,12 @@ libraryDependencies += "org.slf4j" % "slf4j-nop" % "1.7.25"
...
-addSbtPlugin("com.jsuereth"                      % "sbt-pgp"               % "1.0.1")
-addSbtPlugin("org.xerial.sbt"                    % "sbt-sonatype"          % "1.1")
-addSbtPlugin("com.thoughtworks.sbt-api-mappings" % "sbt-api-mappings"      % "0.2.2")
-addSbtPlugin("org.scoverage"                     % "sbt-scoverage"         % "1.5.0")
-addSbtPlugin("org.scala-native"                  % "sbt-scala-native"      % "0.3.1")
-addSbtPlugin("org.scala-native"                  % "sbt-scalajs-crossproject" % "0.2.0")
+addSbtPlugin("com.github.gseitz"                 % "sbt-release"           % "1.0.6")
+addSbtPlugin("com.jsuereth"                      % "sbt-pgp"               % "1.1.0-M1")
+addSbtPlugin("org.xerial.sbt"                    % "sbt-sonatype"          % "2.0")
+addSbtPlugin("com.thoughtworks.sbt-api-mappings" % "sbt-api-mappings"      % "0.2.2") //https://github.com/ThoughtWorksInc/sbt-api-mappings/pull/7
```
Comment (larsrh) [inline project/plugins.sbt]: `sbt-api-mappings` 2.0.0 is out.

**Comment by larsrh** [comment]
@SystemFw I took the liberty to resolve the merge conflict.

**Comment by milessabin** [comment]
Have things moved along enough to finish this off now?

**Comment by larsrh** [comment]
> Have things moved along enough to finish this off now?

Not quite. As far as I can tell, there are four plugins missing; one of them working on it. The others I have on my radar and will ping here if they're ready.

**Comment by SystemFw** [comment]
I think it's 3 left now (I've updated sbt-api-mappings). The two scala native ones have issues or PR almost ready, I'm subscribed and will update as soon as they do. scoverage seems to be the odd one out.

**Comment by dwijnand** [comment]
sbt-scoverage 1.5.1-SNAPSHOT has been released: https://github.com/scoverage/sbt-scoverage/issues/233#issuecomment-322289579

**Comment by larsrh** [comment]
@SystemFw `sbt-osgi` is still missing, as far as I can tell. I was planning to work on that one.

**Comment by SystemFw** [comment]
@larsrh My bad, I had overlooked that one

File: project/build.properties [hunk 3]
```diff
@@ -1 +1 @@
-sbt.version=0.13.16
+sbt.version=1.0.0-RC2
```
Comment (dwijnand) [inline project/build.properties]: 1.0.2 is out

Comment (SystemFw) [reply project/build.properties]: Thanks, I should definitely upgrade. As a more general update, still blocked on scala-native (last checked last week)

Comment (dwijnand) [reply project/build.properties]: Yeah. Though I see that https://github.com/scala-native/scala-native/issues/915 is now labelled as "hot" so hopefully that means "high priority".

Comment (SystemFw) [reply project/build.properties]: Yeah, seen that :) Hopefully it'll be available soon

**Comment by milessabin** [comment]
I could do with some help getting this over the finish line ...

**Comment by SystemFw** [comment]
Yep, sorry. The scala-native thing took ages and I got distracted, so this slipped through the cracks.

**Comment by larsrh** [comment]
Should be a simple matter of bumping the native plugin to 0.3.6.

**Comment by joroKr21** [comment]
sbt 1.1.0 is out

**Comment by SystemFw** [comment]
@joroKr21 Thanks. I'll try to get this done this week, unless someone beats me to it

**Comment by SystemFw** [comment]
Almost done

**Comment by SystemFw** [comment]
This should be done, as far as I can see, sorry for taking so long :)

 The caveat is that there's loads of things only owners can test (like publishing)

**Comment by milessabin** [comment]
MiMa version bump needed for 1.1.0?

**Comment by dwijnand** [comment]
Yeah, to avoid the bincompat issue in scala-compiler. I'd suggest 0.1.18 (the latest): https://github.com/lightbend/migration-manager/releases.

**Comment by SystemFw** [comment]
ah, I'll take care of that then, thanks

File: build.sbt [hunk 4]
```diff
@@ -300,7 +299,7 @@ lazy val mimaSettings = mimaDefaultSettings ++ Seq(
       val previousSJSBinaryVersion =
         ScalaJSCrossVersion.binaryScalaJSVersion(previousSJSVersion)
       val previousBinaryCrossVersion =
-        CrossVersion.binaryMapped(v => s"sjs${previousSJSBinaryVersion}_$v")
+        CrossVersion.binaryWith("sjs${previousSJSBinaryVersion}_", "")
```
Comment (SystemFw) [inline build.sbt]: missing a string interpolator here, fixing now

**Comment by SystemFw** [comment]
`validate` passes on my machine, but there's a CI error related to scald-native. Zero experience with it so any help in getting this through is appreciated :)

**Comment by dwijnand** [comment]
Here's the info I found on Scala Native and re2: http://scala-native.readthedocs.io/en/v0.3.6/user/setup.html#installing-clang-and-runtime-dependencies

**Comment by SystemFw** [comment]
from the `travis.yml`
```yaml
matrix:
  include:
  - scala: 2.11.12
    env: NATIVE=true
    jdk: oraclejdk8
    sudo: required
    before_install:
    - curl https://raw.githubusercontent.com/scala-native/scala-native/v0.3.6/bin/travis_setup.sh | travis_wait 15 bash -x
```

The script run before install is:
```bash
#!/usr/bin/env bash

# Enable strict mode and fail the script on non-zero exit code,
# unresolved variable or pipe failure.
set -euo pipefail
IFS=$'\n\t'

if [ "$(uname)" == "Darwin" ]; then

    brew update
    brew install sbt
    brew install bdw-gc
    brew link bdw-gc
    brew install jq
    brew install re2
    brew install llvm@4
    export PATH="/usr/local/opt/llvm@4/bin:$PATH"

else

    # Remove libunwind pre-bundled with clang
    sudo find /usr -name "*libunwind*" -delete

    # Install Boehm GC, libunwind
    sudo apt-get install -y -qq \
      clang++-3.9 \
      libgc-dev \
      libunwind8-dev

    # Install re2
    # Starting from Ubuntu 16.04 LTS, it'll be available as http://packages.ubuntu.com/xenial/libre2-dev
    sudo apt-get install -y make
    export CXX=clang++-3.9
    git clone https://code.googlesource.com/re2
    pushd re2
    git checkout 2017-03-01
    make -j4 test
    sudo make install prefix=/usr
    make testinstall prefix=/usr
    popd

fi
```
Which should be installing all that's needed. Looks like there has been a connection error at the start of the job, not sure if it's related though

**Comment by larsrh** [comment]
> Which should be installing all that's needed. Looks like there has been a connection error at the start of the job, not sure if it's related though

I've restarted the job, let's see.

**Comment by dwijnand** [comment]
> I've restarted the job, let's see.

Sadly still failed. With:

```
$ travis_wait 60 scripts/travis-publish.sh


Still running (5 of 60): scripts/travis-publish.sh
The command scripts/travis-publish.sh exited with 1.
```

**Comment by larsrh** [comment]
This looks like a Travis problem. Earlier in the log:

```
/home/travis/.travis/job_stages: line 169:  4952 Terminated              travis_jigger $! $timeout $cmd
```

**Comment by milessabin** [comment]
I'm seeing a C compile error in the log.

**Comment by dwijnand** [comment]
I think the issue is the base Ubuntu version. The build says:

> This job ran on our __Trusty__ environment, which has been updated on Tuesday, December 12th, 2017. Read all about this [on our blog](https://blog.travis-ci.com/2017-12-12-new-trusty-images-q4-launch) and take note that you can add `group: deprecated-2017Q4` in your .travis.yml file to use the previous image version.

And http://scala-native.readthedocs.io/en/v0.3.6/user/setup.html#installing-clang-and-runtime-dependencies mentions Ubuntu 16.04.

* Trusty is Ubuntu 14.04 and is an LTS release
* Ubuntu 16.04 is the latest LTS release


So perhaps for Scala Native and re2 reasons CI needs to upgraded to Ubuntu 16.04.

**Comment by milessabin** [comment]
I've asked for help in the scala-native gitter channel ...

**Comment by milessabin** [comment]
Fix here from @densh: https://github.com/scala-native/scala-native/pull/1141#pullrequestreview-87602224.

For now we need to update the URL to fetch the script from to,
```
https://raw.githubusercontent.com/densh/scala-native/da72f8f073553a653aba742b48af8523fc7de380/bin/travis_setup.sh
```
The long term solution is here: https://github.com/scala-native/scala-native/issues/1139.

**Comment by milessabin** [comment]
That's been merged now, so the URL is,

```
https://github.com/scala-native/scala-native/blob/master/bin/travis_setup.sh
```

File: .travis.yml [hunk 5]
```diff
@@ -39,4 +39,4 @@ matrix:
     jdk: oraclejdk8
     sudo: required
     before_install:
-    - curl https://raw.githubusercontent.com/scala-native/scala-native/v0.3.6/bin/travis_setup.sh | travis_wait 15 bash -x
+    - curl https://github.com/scala-native/scala-native/blob/master/bin/travis_setup.sh | travis_wait 15 bash -x
```
Comment (densh) [inline .travis.yml]: This is going to download html content for that file, not the raw data. Naturally, bash wouldn't be able to run that. Hence, the failure.

Comment (densh) [reply .travis.yml]: Correct link: https://raw.githubusercontent.com/scala-native/scala-native/master/bin/travis_setup.sh

Comment (SystemFw) [reply .travis.yml]: I've corrected the link, but it seems to fail regardless...
(Thx for you help btw)

**Comment by milessabin** [comment]
@densh still no joy ... what are we doing wrong here?

Nb. we have native builds working for SBT 0.13.x, so it's the version bumps from there on which seem to be problematic.

File: project/plugins.sbt [hunk 6]
```diff
@@ -1,15 +1,15 @@
...
 addSbtPlugin("com.eed3si9n"                      % "sbt-buildinfo"         % "0.7.0")
-addSbtPlugin("com.typesafe.sbt"                  % "sbt-git"               % "0.9.2")
+addSbtPlugin("com.typesafe.sbt"                  % "sbt-git"               % "0.9.3")
 addSbtPlugin("org.scala-js"                      % "sbt-scalajs"           % "0.6.21")
-addSbtPlugin("com.github.gseitz"                 % "sbt-release"           % "1.0.4")
-addSbtPlugin("com.jsuereth"                      % "sbt-pgp"               % "1.0.1")
-addSbtPlugin("org.xerial.sbt"                    % "sbt-sonatype"          % "1.1")
-addSbtPlugin("com.thoughtworks.sbt-api-mappings" % "sbt-api-mappings"      % "0.2.2")
-addSbtPlugin("org.scoverage"                     % "sbt-scoverage"         % "1.5.0")
+addSbtPlugin("com.github.gseitz"                 % "sbt-release"           % "1.0.6")
```
Comment (xuwei-k) [inline project/plugins.sbt]: sbt-release 1.0.6 does not work with sbt 1.1.0. please update to `1.0.7`
- https://github.com/sbt/sbt/issues/3792
- https://github.com/sbt/sbt-release/commit/66728b59bbf9d9463b69928bc16aee514c94f1d0

File: .travis.yml [hunk 7]
```diff
@@ -39,4 +39,4 @@ matrix:
     jdk: oraclejdk8
     sudo: required
     before_install:
-    - curl https://raw.githubusercontent.com/scala-native/scala-native/v0.3.6/bin/travis_setup.sh | travis_wait 15 bash -x
+    - curl  https://raw.githubusercontent.com/scala-native/scala-native/master/bin/travis_setup.sh | travis_wait 15 bash -x
```
Comment (xuwei-k) [inline .travis.yml]: Maybe it works if remove `travis_wait 15`

**Comment by xuwei-k** [comment]
https://github.com/milessabin/shapeless/pull/813

**Comment by larsrh** [comment]
Closing this in favour of #813, which is pretty much this pull request with some additional changes to make scala-native work. Thanks for your work, @SystemFw et.al.!

**Pull request closed without merging**